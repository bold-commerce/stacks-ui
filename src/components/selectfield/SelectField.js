import React, { Component } from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';

import Field from '../field/Field';
import Select from '../select/Select';
import { generateUniqueId } from '../../utils/generateUniqueId';

import './SelectField.css';

class SelectField extends Component {
  constructor(props) {
    super(props);
    this.forId = `stxField${generateUniqueId()}`;
  }

  render() {
    const { id, label, className, disabled, readOnly, messageType, messageText, helpText, ...otherProps } = this.props;
    const classNames = cn([
      'stx-field--with-select',
      className,
    ]);

    return (
      <Field
        label={label}
        htmlFor={id || this.forId}
        helpText={helpText}
        className={classNames}
        messageType={messageType}
        messageText={messageText}
        disabled={disabled}
        readOnly={readOnly}
      >
        <Select
          {...otherProps}
          id={id || this.forId}
          messageType={messageType}
          disabled={disabled}
          readOnly={readOnly}
        />
      </Field>
    );
  }
}

SelectField.defaultProps = {};

SelectField.propTypes = {
  /** The name or the title of the field */
  label: PropTypes.node,
  /** Helpful text that describes how to use the field */
  helpText: PropTypes.node,
  /** HTML ID attribute for the select element */
  id: PropTypes.string,
  /** Name for the select component */
  name: PropTypes.string,
  /** The value for the select component */
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** Optional className for the select component */
  className: PropTypes.string,
  /** Placeholder for the select component */
  placeholder: PropTypes.string,
  /** The text for the message */
  messageText: PropTypes.string,

  /** The type of message to display */
  messageType: PropTypes.oneOf(['alert', 'warning', 'success', '']),
  /** The order in which fields should gain focus when pressing Tab */
  tabIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

  /** Whether or not the field is disabled */
  disabled: PropTypes.bool,
  /** Whether or not the field is loading */
  loading: PropTypes.bool,
  /** Whether or not the field is read only */
  readOnly: PropTypes.bool,

  /** An array of objects that can be used to build the option set */
  options: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  })),

  /** Event fires when the field gains focus */
  onFocus: PropTypes.func,
  /** Event fires when the field loses focus */
  onBlur: PropTypes.func,
  /** Event fires when the field's value changes */
  onChange: PropTypes.func,
  /** Child elements */
  children: PropTypes.node,
};

export default SelectField;
