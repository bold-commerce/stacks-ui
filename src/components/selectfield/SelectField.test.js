import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import SelectField from './SelectField';

describe('SelectField', () => {
  const onFocus = () => 'onFocus';
  const onBlur = () => 'onBlur';
  const onChange = () => 'onChange';

  test('SelectField renders with basic props', () => {
    const output = mount(<SelectField
      className='the-class'
      onFocus={onFocus}
      onBlur={onBlur}
      onChange={onChange}
      tabIndex='1'
      placeholder='Select something'
      placeholderValue='placeholder value'
      ariaLive='assertive'
    >
      <option value='one'>One</option>
      <option value='two'>Two</option>
      <option value='three'>Three</option>
    </SelectField>);

    expect(toJson(output)).toMatchSnapshot();
  });
});
