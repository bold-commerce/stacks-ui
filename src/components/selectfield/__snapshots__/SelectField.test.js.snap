// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField SelectField renders with basic props 1`] = `
<SelectField
  ariaLive="assertive"
  className="the-class"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="Select something"
  placeholderValue="placeholder value"
  tabIndex="1"
>
  <Field
    ariaLive="assertive"
    className="stx-field--with-select the-class"
    htmlFor="stxField1"
    messageId="stxFieldMessage2"
  >
    <div
      className="stx-field stx-field--with-select the-class"
    >
      <Select
        aria-describedby="stxFieldMessage2"
        id="stxField1"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Select something"
        placeholderValue="placeholder value"
        tabIndex="1"
      >
        <div
          className="stx-select stx-select--with-placeholder"
        >
          <select
            aria-describedby="stxFieldMessage2"
            className="stx-select__select-element"
            id="stxField1"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            tabIndex="1"
            value="placeholder value"
          >
            <option
              hidden={true}
              value="placeholder value"
            >
              Select something
            </option>
            <option
              value="one"
            >
              One
            </option>
            <option
              value="two"
            >
              Two
            </option>
            <option
              value="three"
            >
              Three
            </option>
          </select>
          <span
            aria-hidden={true}
            className="stx-select__icon"
          >
            <svg
              className="stx-select-icon__icon"
              viewBox="0 0 24 24"
            >
              <path
                className="stx-select-icon__path"
                d="M3.75261265,6.34153452 C3.38894463,5.92588561 2.75718344,5.88374677 2.34153452,6.24741478 C1.92588561,6.6110828 1.88374677,7.24284399 2.24741478,7.65849291 L11.2524148,18.2364929 C11.6508167,18.6918404 12.3591606,18.6918563 12.757583,18.2365268 L21.763583,7.6585268 C22.1272697,7.24289426 22.0851593,6.61113117 21.6695268,6.24744443 C21.2538943,5.8837577 20.6221312,5.9258681 20.2584444,6.34150064 L12.0050479,16.0594019 L3.75261265,6.34153452 Z"
                fill="#6B6B6B"
              />
            </svg>
          </span>
        </div>
      </Select>
    </div>
  </Field>
</SelectField>
`;
